File: Information-Security-Lab-master/EX - 3/RSA.java
import java.util.*;
import java.math.*;

public class RSA
{
	public boolean gcd_verify(BigInteger e, BigInteger phi)
	{
		BigInteger gcd_value = e.gcd(phi);
		System.out.println(gcd_value);
		if(gcd_value.compareTo(BigInteger.valueOf(1))==0)
			return true;
		else
			return false;
	}

	public BigInteger encryptkey(BigInteger phi)
	{
		BigInteger e;
		Scanner input = new Scanner(System.in);
		while(true)
		{
			System.out.println("Enter the value of e: ");
			e = input.nextBigInteger();
			if(gcd_verify(e,phi)==true)
			{
				System.out.println("e value is valid!");
				break;
			}
			else
				continue;
		}

		System.out.println("e:  "+e);

		return e;
	}

	public BigInteger decryptkey(BigInteger e, BigInteger phi)
	{
		BigInteger a=phi,b=e,q,r,t,t1,t2;
		t1=BigInteger.valueOf(0);
		t2=BigInteger.valueOf(1);
		while(!(b.equals(BigInteger.valueOf(0))))
		{
			q=a.divide(b);
			r=a.mod(b);
			t=t1.subtract(q.multiply(t2));
			a=b;
			b=r;
			t1=t2;
			t2=t;
		}
		if(t1.compareTo(BigInteger.valueOf(0))==-1)
		{
			t1=t1.add(phi);
		}
		System.out.println("d:  "+t1);
		return t1;
	}


	public BigInteger encrypt(BigInteger plain, BigInteger e, BigInteger n)
	{
		return plain.modPow(e, n);
	}

	public BigInteger decrypt(BigInteger cipher, BigInteger d, BigInteger n)
	{
		return cipher.modPow(d, n);
	}
	
	public static void main(String[] args)
	{
		Scanner input = new Scanner(System.in);
		RSA rsa = new RSA();

		String  alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

		while(true)
		{
			System.out.println("Enter your choice:\n1-Key generation\n2-Encryption\n3-Decryption\n4-Exit");
			BigInteger n, e, phi, d;
			int choice = input.nextInt();
			if(choice==4)
			{
				break;
			}
			else
			{
				switch(choice)
				{
					case 1:
					{
						System.out.println("Enter the prime number (p):");
						BigInteger p = input.nextBigInteger();
						System.out.println("Enter the prime numer (q):");
						BigInteger q = input.nextBigInteger();
						
						n = p.multiply(q);
						
						BigInteger one = BigInteger.valueOf(1);
					
						phi = (p.subtract(one)).multiply((q.subtract(one)));

						System.out.println("n:  " + n);
						System.out.println("phi:  " + phi);
						System.out.println();

						e = rsa.encryptkey(phi);
						d = rsa.decryptkey(e, phi);
						
						break;
					}
					case 2:
					{
						System.out.println("Enter the plain text:");
						String pt = input.next().toUpperCase();
						System.out.println("Enter the multiplication of two prime numbers (n):");
						n = input.nextBigInteger();
						System.out.println("Enter the encryption key (e):");
						e = input.nextBigInteger();
						System.out.println("Cipher:");
						for(int i=0; i<pt.length(); i++)
						{
							BigInteger plain = BigInteger.valueOf(alpha.indexOf(pt.charAt(i)));
							System.out.print(rsa.encrypt(plain, e, n) + "\n");
						}
						break;
					}
					case 3:
					{
						System.out.println("Enter the number of characters:");
						int count = input.nextInt();

						BigInteger[] num_array = new BigInteger[count];
                        System.out.println("Enter the CT:");
						for(int i=0; i<count; i++)
						{
							num_array[i] = input.nextBigInteger();
						}

						System.out.println("Enter the multiplication of two prime numbers (n):");
						n = input.nextBigInteger();
						System.out.println("Enter the decryption key (d)");
						d = input.nextBigInteger();
						System.out.println("plain text: ");
						for(int i=0;i<count; i++)
						{
							System.out.print(alpha.charAt(rsa.decrypt(num_array[i], d, n).intValue()));
						}
						System.out.println();
						break;
					}
					default:
						System.out.println("Enter the valid input!");

				}
			}

		
		}
	}
}

File: practice/rsa.py
pubkey,pvtkey=[],[]
def mul_inv(a,b):
    t1=0
    t2=1
    x=a
    while b!=0:
        q=a//b
        r=a%b
        a=b
        b=r
        # inv
        t=t1-q*t2
        t1=t2
        t2=t  
    if a==1 :
        return t1 % x
    else:
        print("GCD Not Equal to 1")
def NumEnc(p):
    alpha="abcdefghijklmnopqrstuvwxyz"
    temp=[]
    for i in p:
        temp.append(alpha.index(i))
    return temp
def Encrypt(pt,pubkey) :
    msg=NumEnc(pt) 
    ct=[]  
    for i in msg:
        j=pow(i,pubkey[0],pubkey[1]) 
        ct.append(j) 
    return ct   
def Decrypt(ct,pvtkey):
    pt=[]
    plaintext=""
    for i in ct:
        j=pow(i,pvtkey[0],pvtkey[1]) 
        pt.append(j)
    alpha="abcdefghijklmnopqrstuvwxyz"
    for i in pt:
        plaintext+=alpha[i]
    return plaintext                                                                                                                     
def Gen_key(e,phy):
    d=mul_inv(phy,e)
    pubkey.append(e)
    pvtkey.append(d)
    pubkey.append(n)
    pvtkey.append(n)
    print("Public Key:",pubkey)
    print("Private Key:",pvtkey)
while(True):
    print("******RSA******")
    print("1.Key Generation\n2.Encryption\n3.Decryption\n4.Exit")
    choice= int(input("Enter choice:"))
    if choice==1:
        p=int(input("Enter a large prime number:"))
        q=int(input("Enter a large prime number:"))
        e=int(input("Enter public Key:"))
        n=p*q
        phy=(p-1)*(q-1)
        Gen_key(e,phy)
    elif choice==2:
        pt=input("Enter Message:")
        pbkey=list(map(int,input("Enter Public Key pair:").split(" ")))
        ct=Encrypt(pt,pbkey)
        print("Ciphe text numeric Values is",ct)
    elif choice==3:
        ct=list(map(int,input("Enter Cipher text numeric Values").split(" ")))
        ptkey=list(map(int,input("Enter Private  Key pair:").split(" ")))
        print("Plaintext is:",Decrypt(ct,ptkey))
    elif choice==4:
        break
    else:
        print("Enter a valid Choice")
